Automatically save log files at the end of managed runs

ealt opened the pull request:
Resolves #141

ealt committed:
Automatically save log files at the end of managed runs

ealt committed:
Include base logger files

ealt committed:
Include path in error message

ealt committed:
Improve error handling during cleanup

loren-ac reviewed (approved):
As far as I can tell this looks good. One thing I feel confused about is how ray and hydra interact with logging and if there are any weird edge cases in the multirun setting but if this is something you aren't concerned about then I'm good with it!

adamimos commented:
## Code Review

Found 2 issues:

### 1. `get_log_files()` misses the root logger's handlers

The function iterates over `logging.Logger.manager.loggerDict`, but Python's root logger is **not** stored in `loggerDict` - it's a special logger accessed via `logging.getLogger()` (no arguments) or `logging.root`. This means log files attached to the root logger (like the `fileHandler` configured on `logger_root` in logging.ini) will not be discovered, saved as artifacts, or cleaned up.

The irony: `add_handlers_to_existing_loggers()` in the same file correctly accesses the root logger with `root_logger = logging.getLogger()`, but this pattern wasn't applied to `get_log_files()`.

**Fix:** Add the root logger's handlers to the collection:
```python
def get_log_files() -> list[str]:
    log_files = []
    # Check root logger (not in loggerDict)
    root_logger = logging.getLogger()
    log_files.extend(
        [h.baseFilename for h in root_logger.handlers if isinstance(h, logging.FileHandler)]
    )
    # Check named loggers
    for logger_name in logging.Logger.manager.loggerDict:
        ...
```

https://github.com/Astera-org/simplexity/blob/cf7d355b4fef2f2628414eca1baa1b282c1631e2/simplexity/logger.py#L69-L78

---

### 2. `os.remove(log_file)` lacks error handling

In `_cleanup()`, `os.remove(log_file)` is called without checking if the file exists or catching exceptions. The `baseFilename` from FileHandler represents the *configured* path, not a guarantee the file was created. If no log messages were written (e.g., due to log level filtering) or if cleanup runs after an early failure, `FileNotFoundError` will crash the cleanup phase.

**Fix:** Use `pathlib.Path(log_file).unlink(missing_ok=True)` (Python 3.8+) or wrap in try/except.

https://github.com/Astera-org/simplexity/blob/cf7d355b4fef2f2628414eca1baa1b282c1631e2/simplexity/run_management/run_management.py#L694-L697

---

### Re: loren-ac's comment on Ray/Hydra multirun edge cases

Good instinct! In multirun settings, each worker process will have its own logging setup. Key considerations:
- Each worker calling `_setup_python_logging()` will configure its own handlers independently
- If multiple workers write to the same log file path (e.g., all writing to `training.log`), you'll get interleaved/corrupted output
- The cleanup phase could have race conditions if workers finish at different times

If multirun support is desired, the logging.ini should use a unique filename per worker (e.g., including `%(process)d` in the path) or the feature should be disabled for distributed runs. Worth documenting or testing!

---

Checked for bugs and CLAUDE.md compliance. The CLAUDE.md violations flagged (inline comments, missing docstring) were false positives upon closer inspection.

adamimos commented:
@ealt could you look at Claude's review and determine if any of these issues make sense/are important?

ealt committed:
Test base logger files

ealt committed:
Exception handling

ealt committed:
Add context managers to preserve logging state in tests

- Introduced `preserve_logging_state` context manager to save and restore the global logging configuration during tests.
- Added `with_preserved_logging_state` decorator to apply the context manager to test functions, ensuring consistent logging behavior.
- Updated test functions to utilize the new decorator for improved logging state management.

ealt committed:
Merge remote-tracking branch 'origin/dev' into logs

ealt committed:
Fix test

ealt commented:
> @ealt could you look at Claude's review and determine if any of these issues make sense/are important?

I resolved those issues (Code Rabbit found them for me too!)

adamimos commented:
Nice! I'm currently running a run with the log code, assuming it puts the log on mlflow I will accept the PR and merge.

ealt committed:
Add log file management functions to logger

- Introduced `remove_file_handlers`, `remove_log_file`, and `remove_log_files` functions to enhance log file management.
- Updated `_cleanup` function to utilize `remove_log_files` for better cleanup of log files.
- Enhanced tests to cover new log file removal functionalities, ensuring proper handling of file handlers and log files in various scenarios.
